RecoMuon/TrackerSeedGenerator/plugins/TSGForOIDNN.cc

```diff
@@ -0,0 +1,949 @@
+    hitlessSeedsMadeIP = 0;
+    hitlessSeedsMadeMuS = 0;
+    hitSeedsMade = 0;
+    hitDoubletSeedsMade = 0;
+
+    // BARREL
+    if (absL2muonEta < maxEtaForTOB_) {
+      layerCount = 0;
+      for (auto it = tob.rbegin(); it != tob.rend(); ++it) {
+        LogTrace("TSGForOIDNN") << "TSGForOIDNN::produce: looping in TOB layer " << layerCount;
+
+        if (hitlessSeedsMadeIP < maxHitlessSeedsIP && numSeedsMade < maxSeeds_)
+          makeSeedsWithoutHits(**it,
+                               tsosAtIP,
+                               *(propagatorAlong.get()),
+                               estimator,
+                               errorSFHitless,
+                               hitlessSeedsMadeIP,
+                               numSeedsMade,
+                               out);
+
+        if (outerTkStateInside.isValid() && outerTkStateOutside.isValid() && hitlessSeedsMadeMuS < maxHitlessSeedsMuS &&
+            numSeedsMade < maxSeeds_)
+          makeSeedsWithoutHits(**it,
+                               outerTkStateOutside,
+                               *(propagatorOpposite.get()),
+                               estimator,
+                               errorSFHitless,
+                               hitlessSeedsMadeMuS,
+                               numSeedsMade,
+                               out);
+
+        if (hitSeedsMade < maxHitSeeds && numSeedsMade < maxSeeds_)
+          makeSeedsFromHits(**it,
+                            tsosAtIP,
+                            *(propagatorAlong.get()),
+                            estimator,
+                            measurementTracker,
+                            hitSeedsMade,
+                            numSeedsMade,
+                            layerCount,
+                            out);
+
+        if (hitDoubletSeedsMade < maxHitDoubletSeeds && numSeedsMade < maxSeeds_)
+          makeSeedsFromHitDoublets(**it,
+                                   tsosAtIP,
+                                   *(propagatorAlong.get()),
+                                   estimator,
+                                   measurementTracker,
+                                   navSchool,
+                                   hitDoubletSeedsMade,
+                                   numSeedsMade,
+                                   layerCount,
+                                   out);
+      }
+      LogTrace("TSGForOIDNN") << "TSGForOIDNN:::produce: NumSeedsMade = " << numSeedsMade
+                              << " , layerCount = " << layerCount;
+    }
+
+    // Reset number of seeds if in overlap region
+    if (absL2muonEta > minEtaForTEC_ && absL2muonEta < maxEtaForTOB_) {
+      numSeedsMade = 0;
+      hitlessSeedsMadeIP = 0;
+      hitlessSeedsMadeMuS = 0;
+      hitSeedsMade = 0;
+      hitDoubletSeedsMade = 0;
+    }
+
+    // ENDCAP+
+    if (l2muonEta > minEtaForTEC_) {
+      layerCount = 0;
+      for (auto it = tecPositive.rbegin(); it != tecPositive.rend(); ++it) {
```

Comment:
https://github.com/cms-sw/cmssw/pull/35237#discussion_r712607565

> the code inside the loop is repeated 3 times for the barrel, and the two endcaps.
> Please refactor to e.g. a lambda and reuse instead of copy-pasting about 40 lines of code in each case.

Reply:
Done!